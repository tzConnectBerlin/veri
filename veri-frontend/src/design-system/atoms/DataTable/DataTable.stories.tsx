import { Story, Meta } from '@storybook/react/types-6-0';
import { DataTable, DataTableProps } from './DataTable';

export default {
  title: 'Atoms/DataTable',
  component: DataTable,
} as Meta;

const Template: Story<DataTableProps> = args => <DataTable {...args} />;

const SamleData: DataTableProps = {
  header: [
    { field: 'event_name', value: 'Event Name', type: 'String' },
    { field: 'organizer', value: 'Organizer', type: 'String' },
    { field: 'status', value: 'Status', type: 'String', sortable: true },
    {
      field: 'mint_date',
      value: 'Mint Date',
      type: 'String',
    },
  ],
  rows: [
    {
      cols: [
        { field: 'event_name', value: 'Event1', type: 'String' },
        { field: 'organizer', value: 'Organizer1', type: 'String' },
        {
          field: 'status',
          value: 'Draft',
          type: 'String',
          sortable: true,
        },
        {
          field: 'mint_date',
          value: '21 Nov 2022',
          type: 'String',
        },
      ],
      actions: {
        onDelete: () => '',
        onEdit: () => '',
      },
    },
    {
      cols: [
        { field: 'event_name', value: 'Event1', type: 'String' },
        { field: 'organizer', value: 'Organizer1', type: 'String' },
        {
          field: 'status',
          value: 'Draft',
          type: 'String',
          sortable: true,
        },
        {
          field: 'mint_date',
          value: '21 Nov 2022',
          type: 'String',
        },
      ],
      actions: {
        onDelete: () => '',
        onEdit: () => '',
      },
    },
    {
      cols: [
        { field: 'event_name', value: 'Event1', type: 'String' },
        { field: 'organizer', value: 'Organizer1', type: 'String' },
        {
          field: 'status',
          value: 'Draft',
          type: 'String',
          sortable: true,
        },
        {
          field: 'mint_date',
          value: '21 Nov 2022',
          type: 'String',
        },
      ],
      actions: {
        onDelete: () => '',
        onEdit: () => '',
      },
    },
  ],
  hasActions: true,
};

export const NoActions = Template.bind({});
NoActions.args = {
  ...SamleData,
  hasActions: false,
};
export const WithActions = Template.bind({});
WithActions.args = {
  ...SamleData,
};
export const WithTitle = Template.bind({});
WithTitle.args = {
  ...SamleData,
  title: 'All Veris',
};
