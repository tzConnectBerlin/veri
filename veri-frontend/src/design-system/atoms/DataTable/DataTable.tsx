import {
  Box,
  Button,
  ButtonGroup,
  Flex,
  IconButton,
  Table,
  TableCaption,
  TableContainer,
  Tbody,
  Td,
  Th,
  Thead,
  Tr,
} from '@chakra-ui/react';
import { IoMdArrowDropdown } from 'react-icons/io';
import { MdEdit, MdDelete } from 'react-icons/md';
import React from 'react';

interface column {
  field: string;
  isNum?: boolean;
  value: string | number | React.ReactNode;
  sortable?: boolean;
}

interface row {
  cols: column[];
  action?: () => void;
}

export interface DataTableProps {
  title?: string;
  header: column[];
  rows: row[];
}

export const DataTable: React.FC<DataTableProps> = ({
  title,
  header,
  rows,
}) => {
  return (
    <Box>
      <TableContainer>
        <Table variant="simple">
          {title && (
            <TableCaption placement="top" fontSize="2xl">
              {title}
            </TableCaption>
          )}
          <Thead>
            <Tr>
              {header.map(h => (
                <Th key={h.field} isNumeric={h.isNum}>
                  {h.sortable ? (
                    <Flex alignItems="center">
                      {h.value}
                      <IconButton
                        size="xs"
                        variant="ghost"
                        aria-label="sort"
                        icon={<IoMdArrowDropdown />}
                      />
                    </Flex>
                  ) : (
                    h.value
                  )}
                </Th>
              ))}
            </Tr>
          </Thead>
          <Tbody>
            {rows.map((row, i) => (
              <Tr key={i} onClick={row.action}>
                {row.cols.map(col => (
                  <Td
                    key={col.field}
                    isNumeric={col.isNum}
                    borderBottomColor={
                      i === rows.length - 1 ? 'transparent' : 'inherit'
                    }
                  >
                    {col.value}
                  </Td>
                ))}
              </Tr>
            ))}
          </Tbody>
        </Table>
      </TableContainer>
    </Box>
  );
};
