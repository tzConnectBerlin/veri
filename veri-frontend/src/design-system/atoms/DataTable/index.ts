export { DataTable, DataTable as default } from './DataTable';
export type { DataTableProps } from './DataTable';
