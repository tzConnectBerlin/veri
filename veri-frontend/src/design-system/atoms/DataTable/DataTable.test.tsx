import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { DataTable, DataTableProps } from './DataTable';

const TestData: DataTableProps = {
  header: [
    { field: 'event_name', value: 'Event Name' },
    { field: 'organizer', value: 'Organizer' },
    { field: 'status', value: 'Status', sortable: true },
    {
      field: 'mint_date',
      value: 'Mint Date',
    },
  ],
  rows: [
    {
      cols: [
        { field: 'event_name', value: 'Event1' },
        { field: 'organizer', value: 'Organizer1' },
        {
          field: 'status',
          value: 'Draft',
          sortable: true,
        },
        {
          field: 'mint_date',
          value: '21 Nov 2022',
        },
      ],
    },
    {
      cols: [
        { field: 'event_name', value: 'Event1' },
        { field: 'organizer', value: 'Organizer1' },
        {
          field: 'status',
          value: 'Draft',
          sortable: true,
        },
        {
          field: 'mint_date',
          value: '21 Nov 2022',
        },
      ],
    },
    {
      cols: [
        { field: 'event_name', value: 'Event1' },
        { field: 'organizer', value: 'Organizer1' },
        {
          field: 'status',
          value: 'Draft',
          sortable: true,
        },
        {
          field: 'mint_date',
          value: '21 Nov 2022',
        },
      ],
    },
  ],
};

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

describe('Snapshot testing DataTable Component', () => {
  it('renders correctly', () => {
    const Component = renderer.create(<DataTable {...TestData} />).toJSON();
    expect(Component).toMatchSnapshot();
  });
});

describe('Element testing DataTable Component', () => {
  it('render correctly title, cols and rows', async () => {
    const { container } = render(<DataTable {...TestData} title="All Veris" />);
    const title = screen.getByText(/All Veris/i);
    const rows = container.querySelectorAll('tr');
    const cols = container.querySelectorAll('th');

    expect(title).toBeInTheDocument();
    expect(rows.length).toEqual(4);
    expect(cols.length).toEqual(4);
  });

  it('render correctly Fire handleSort Event', async () => {
    const sortStatus = jest.fn();
    render(<DataTable {...TestData} handleSort={sortStatus} />);
    const button = await screen.getByRole('button');
    fireEvent.click(button);
    expect(sortStatus).toBeCalledTimes(1);
  });
});
