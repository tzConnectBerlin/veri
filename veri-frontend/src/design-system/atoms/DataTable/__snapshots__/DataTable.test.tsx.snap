// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing DataTable Component renders correctly 1`] = `
<div
  className="css-0"
>
  <div
    className="chakra-table__container css-zipzvv"
  >
    <table
      className="chakra-table css-0"
    >
      <thead
        className="css-0"
      >
        <tr
          className="css-0"
        >
          <th
            className="css-0"
          >
            Event Name
          </th>
          <th
            className="css-0"
          >
            Organizer
          </th>
          <th
            className="css-0"
          >
            <div
              className="css-70qvj9"
            >
              Status
              <button
                aria-label="sort"
                className="chakra-button css-p7mkev"
                onClick={[Function]}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  fill="currentColor"
                  focusable={false}
                  height="1em"
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M128 192l128 128 128-128z"
                  />
                </svg>
              </button>
            </div>
          </th>
          <th
            className="css-0"
          >
            Mint Date
          </th>
        </tr>
      </thead>
      <tbody
        className="css-0"
      >
        <tr
          className="css-ykqdxu"
          onClick={[Function]}
        >
          <td
            className="css-130zoje"
          >
            Event1
          </td>
          <td
            className="css-130zoje"
          >
            Organizer1
          </td>
          <td
            className="css-130zoje"
          >
            Draft
          </td>
          <td
            className="css-130zoje"
          >
            21 Nov 2022
          </td>
        </tr>
        <tr
          className="css-ykqdxu"
          onClick={[Function]}
        >
          <td
            className="css-130zoje"
          >
            Event1
          </td>
          <td
            className="css-130zoje"
          >
            Organizer1
          </td>
          <td
            className="css-130zoje"
          >
            Draft
          </td>
          <td
            className="css-130zoje"
          >
            21 Nov 2022
          </td>
        </tr>
        <tr
          className="css-ykqdxu"
          onClick={[Function]}
        >
          <td
            className="css-ipz9xt"
          >
            Event1
          </td>
          <td
            className="css-ipz9xt"
          >
            Organizer1
          </td>
          <td
            className="css-ipz9xt"
          >
            Draft
          </td>
          <td
            className="css-ipz9xt"
          >
            21 Nov 2022
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
